### PR TITLE
Add emulated step

### PR DESCRIPTION
**Pull request by KSGulin**
In #152 `steps_per_epoch` was updated to scale with the gradient accumulation steps. This was a mistake, as the steps_per_epoch is generally not constant in yolov5. During training there is a warmup phase during which the gradient accumulation steps number scales up from 1 to its intended value. Instead, an emulated step should be called during each "non-update" step, which steps forward all SparseML modifiers but not the wrapped scaler/optimizer.

Test Plan
`python train.py --cfg yolov5s.yaml --sparsification-recipe yolov5_quant_recipe.yaml --gradient-accum-steps N`
where N was ran with each of [1, 2, 3] and each time was manually confirmed to follow the expected update pattern activate quantization at the correct step. The recipe used was:

```
# General variables
num_epochs: 6
init_lr: 1.e-6
final_lr: 2.e-7

# Quantization variables
observer_freeze_epoch: 1
bn_freeze_epoch: 1

training_modifiers:
  - !EpochRangeModifier
    start_epoch: 0
    end_epoch: eval(num_epochs)
    
  - !LearningRateFunctionModifier
    start_epoch: 0
    end_epoch: eval(num_epochs)
    lr_func: cosine
    init_lr: eval(init_lr)
    final_lr: eval(final_lr)
       
quantization_modifiers:
  - !QuantizationModifier
    start_epoch: 1.0
    submodules:
      - model
    custom_quantizable_module_types: ['SiLU']
    exclude_module_types: ['SiLU']
    quantize_conv_activations: False
    disable_quantization_observer_epoch: eval(observer_freeze_epoch)
    freeze_bn_stats_epoch: eval(bn_freeze_epoch)
```

### CODE REVIEW

**Review by bfineran — APPROVED** [review]
Nice fix, wonder if we can generalize

**Comment by KSGulin** [comment]
@bfineran agreed this feels like a patch fix, but with `steps_per_epoch `not being constant across epochs and the SparseML modifier manager currently assuming that it is, I couldn't think of a better integration-side solution

**Comment by corey-nm** [comment]
@KSGulin any ideas how we could make an automatic test for integrations based off of your manual test? these details are subtle for each integration and its hard to verify just looking at code

**Comment by KSGulin** [comment]
@corey-nm good question. That's tough. The integrations and SparseML are keeping separate clocks and the integrations are being run as sub processes, so there's not a direct way to make a comparison between their time keeping. Maybe we could design a recipe and command that will fail in all but perfectly in-sync training. But this wouldn't be a generic solution, would need to be hand-tailored for each integration.

But.. the really simple answer is we need dedicated servers for integration tests where we can test full runs and compare based on metrics 😄
